File: Week_11/841. 钥匙和房间/841.钥匙和房间.js
/*
 * @lc app=leetcode.cn id=841 lang=javascript
 *
 * [841] 钥匙和房间
 */

// @lc code=start

var canVisitAllRooms = function(rooms, key = 0, keys = new Set()) {

    if (keys.has(key)) return;

    keys.add(key);

    rooms[key].forEach(k => canVisitAllRooms(rooms, k, keys));

    return keys.size == rooms.length;

};
/**
 * bfs
 * @param {number[][]} rooms
 * @return {boolean}
 */
// var canVisitAllRooms = function(rooms) {
//     let set = new Set();
//     let queue = [];
//     queue.push(0);
//     set.add(0);
//     while (queue.length) {
//         let nextRooms = rooms[queue.shift()];
//         for (let i = 0; i < nextRooms.length; i++) {
//             let next = nextRooms[i];
//             if (!set.has(next)) {
//                 queue.push(next);
//                 set.add(next);
//             }
//         }
//     }
//     if (rooms.length == set.size) {
//         return true;
//     } else {
//         return false;
//     }
// };




// /**
//  * dfs
//  * @param {number[][]} rooms
//  * @return {boolean}
//  */
// var canVisitAllRooms = function (rooms) {
//     let set = new Set();
//     dfs(rooms, 0, set);
//     if (rooms.length == set.size) {
//         return true;
//     } else {
//         return false;
//     }
// };

// function dfs(rooms, room, set) {
//     if (set.has(room)) {
//         return;
//     }
//     set.add(room);
//     for (let item of rooms[room]) {
//         dfs(rooms, item, set);
//     }
// }
// @lc code=end

File: Week_11/841. 钥匙和房间/841.钥匙和房间.java
/*
 * @lc app=leetcode.cn id=841 lang=java
 *
 * [841] 钥匙和房间
 */

// @lc code=start
class Solution {
    /*
     * bfs
     * 从0号开始，统计已经遍历的个数，对比一下。
     */
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Set<Integer> visited = new HashSet<>();
        Deque<Integer> queue = new LinkedList<>();
        queue.addLast(0);
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i=0;i<size;i++){
                int room = queue.removeFirst();
                if(!visited.contains(room)){
                    visited.add(room);
                    for(Integer tempRoom: rooms.get(room)){
                        queue.addLast(tempRoom);
                    }
                }
            }
        } 
        if(visited.size()==rooms.size()){
            return true;
        }else{
            return false;
        }
    }
}



// class Solution {
//     /**
//      * dfs
//      * 从0号开始，统计已经遍历的个数，对比一下。
//      */
//     public boolean canVisitAllRooms(List<List<Integer>> rooms) {
//         Set<Integer> visited = new HashSet<>();
//         dfs(rooms,0,visited);
//         if(visited.size()==rooms.size()){
//             return true;
//         }else{
//             return false;
//         }
//     }

//     private void dfs(List<List<Integer>> rooms,int room,Set<Integer> visited){
//         if(visited.contains(room)){
//             return;
//         }
//         visited.add(room);
//         for(Integer next: rooms.get(room)){
//             dfs(rooms,next,visited);
//         }
//     }
// }
// @lc code=end

